wallet/src/de/schildbach/wallet/ui/transactions/TxResourceMapper.java

```diff
@@ -68,8 +73,15 @@ public int getTransactionTypeName(@NonNull Transaction tx, @NonNull TransactionB
                     }
 
                     if (dashPayWallet != null && CreditFundingTransaction.isCreditFundingTransaction(tx)) {
-                        CreditFundingTransaction cftx = dashPayWallet.getCreditFundingTransaction(tx);
-                        switch (dashPayWallet.getAuthenticationKeyType(cftx.getCreditBurnPublicKeyId().getBytes())) {
+                        AuthenticationGroupExtension authExtension =
+                                (AuthenticationGroupExtension) dashPayWallet.addOrGetExistingExtension(
+                                        // TODO: do we need to call addKeyChains here?
+                                        new AuthenticationGroupExtension(dashPayWallet.getParams())
+                                );
+                        CreditFundingTransaction cftx = authExtension.getCreditFundingTransaction(tx);
+                        // TODO: getKeyChainGroup and getKeyChainType are protected. Any other way to get the type?
+                        AuthenticationKeyChainGroup group = ((AuthenticationKeyChainGroup)authExtension.getKeyChainGroup());
+                        switch (group.getKeyChainType(cftx.getCreditBurnPublicKeyId().getBytes())) {
```

Comment:
if `CreditFundingTransaction.isCreditFundingTransaction(tx)` is `true`, that means that the keychains were already added during as part of username creation.  We don't need to call `addKeyChains` here.

The only place to call `addKeyChains` is during username creation.

Reply:
I made a PR for this:
https://github.com/dashpay/dashj/pull/217

It is part of the CoinJoin branch, which we will be using in the next PR after this one.

You should be able to change the dashj version to `19.1-CJ-SNAPSHOT` and hopefully nothing is missing or broken.  Otherwise, we can make the change change to the master branch of `dashj`

Reply:
@HashEngineering another change might be needed. See the PR